### PR TITLE
Updated progress monitoring

### PR DESCRIPTION
**Pull request by ashesfall**

### CODE REVIEW

File: flowtree/agents/src/main/java/io/flowtree/jobs/AgentProcessRunner.java [hunk 1]
```diff
@@ -194,28 +235,65 @@ private static Result runInTmux(ProcessBuilder pb,
...
+                                   AtomicBoolean loopKilled,
+                                   ConsoleFeatures logger) throws IOException {
+        AgentProgressMonitor progress =
+                loopSignatureExtractor != null ? new AgentProgressMonitor() : null;
+        String line;
+        while ((line = reader.readLine()) != null) {
+            lastOutputAt.set(System.currentTimeMillis());
+            out.append(line).append("\n");
+            logger.log(line);
+            if (progress != null && progress.observe(loopSignatureExtractor.apply(line))) {
```
[the remark this answers is not in the document]
Comment (ashesfall) [reply flowtree/agents/src/main/java/io/flowtree/jobs/AgentProcessRunner.java]: Fixed. Wrapped the `loopSignatureExtractor.apply(line)` call in a try/catch. Exceptions from malformed or unexpected output now produce a null signature (non-action), so the read loop continues and the subprocess is not left running.

File: flowtree/agents/src/main/java/io/flowtree/jobs/agent/OpencodeOutputParser.java [hunk 2]
```diff
@@ -213,6 +248,47 @@ static AgentRunResult parse(String rawOutput,
...
+            return null;
+        }
+        String tool = JsonFieldExtractor.getTextOrNull(part, "tool");
+        if (tool == null || tool.isEmpty()) {
+            return null;
+        }
+        JsonNode input = part.path("state").path("input");
+        String inputText = input.isMissingNode() || input.isNull() ? "" : input.toString();
+        return "tool:" + tool + ":" + Integer.toHexString(inputText.hashCode());
+    }
```
[the remark this answers is not in the document]
Comment (ashesfall) [reply flowtree/agents/src/main/java/io/flowtree/jobs/agent/OpencodeOutputParser.java:322]: Fixed. Replaced `String.hashCode()` with the first 16 hex characters of a SHA-256 digest computed via `MessageDigest.getInstance("SHA-256")`. Added a fallback to hashCode if SHA-256 is unavailable (should never happen with standard JDKs). The digest is computed on the UTF-8 bytes of the full JSON input string.

File: flowtree/agents/src/test/java/io/flowtree/jobs/AgentProgressMonitorTest.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+    /** Repeats that fall outside the sliding window do not accumulate into a false positive. */
+    @Test(timeout = 5000)
+    public void windowEvictionPreventsStaleCounting() {
+        AgentProgressMonitor monitor = new AgentProgressMonitor(3, 5, 4);
+        // One early "x", then enough distinct signatures to evict it from the window.
+        assertFalse(monitor.observe("x"));
+        for (String sig : new String[] {"a", "b", "c", "d", "e"}) {
+            assertFalse(monitor.observe(sig));
+        }
```
[the remark this answers is not in the document]
Comment (ashesfall) [reply flowtree/agents/src/test/java/io/flowtree/jobs/AgentProgressMonitorTest.java:66]: Added `noFalsePositiveOnSelfEviction()` which exercises the exact edge case: a full window where the oldest entry is the same signature being re-observed. The test asserts that after observing "same" (which evicts the prior "same" from the window) the count is correctly 1, no loop is reported, and `getOffendingSignature()` returns null.

File: flowtree/agents/src/main/java/io/flowtree/jobs/AgentProgressMonitor.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    boolean observe(String signature) {
+        if (signature == null) {
+            return false;
+        }
+        totalActions++;
+        window.addLast(signature);
+        int count = counts.merge(signature, 1, Integer::sum);
+        if (window.size() > windowSize) {
+            String evicted = window.removeFirst();
+            if (counts.merge(evicted, -1, Integer::sum) == 0) {
```
[the remark this answers is not in the document]
Comment (ashesfall) [reply flowtree/agents/src/main/java/io/flowtree/jobs/AgentProgressMonitor.java:123]: Fixed. The local `count` variable is now captured *before* adding to the window and counts map, and the post-eviction comparison reads from `counts.getOrDefault(signature, 0)` after the eviction block runs. This ensures that when the same signature is evicted and re-observed in the same call, the window count reflects the post-eviction state (i.e. the just-evicted entry is decremented before checking against `repeatThreshold`).
